Implement DeviceSerialized to smoothly move data to host and back

mrocklin opened the pull request:
So some issues I ran into:

It looks like even though we never put host-like things into `self.device_buffer` they can still appear to be there because that object looks at it's `.slow` attribute, which is `self.host_buffer`.  So when we put something into `self.host_buffer` it can appear to also be in `self.device_buffer`.  To address this I've taken the somewhat kludgy (though also direct) approach of maintaining a separate set of all the things we've placed into or removed from `self.device_buffer`.  This seems to do the trick, and I think it simplifies more than it harms.

I also ran into issues with reading and writing the "frames" that we were getting from `numba.cuda.as_gpu_array(...).copy_to_host()`.  This was creating numpy arrays rather than proper byte-like objects.  We were then writing these numpy arrays to disk without recording their metadata, which resulted in a loss of information.  I've invoked a generic serialization solution for `DeviceSerialized` which should traverse through its attributes and handle things sensibly.  We could improve this a bit by choosing to traverse only through the `.parts/frames` attribute, but I'll leave that for future work.

I've also renamed the `.frames` attribute to `.parts` to signal that what we're storing here in this attribute is more complex than just bytestrings.

mrocklin committed:
Implement DeviceSerialized to smoothly move data to host and back

mrocklin commented:
This runs at about 1.5GB/s, which doesn't seem great

```python
import cupy
x = cupy.arange(10000000)

from dask_cuda.device_host_file import device_to_host, serialize

%time _ = device_to_host(x)
# CPU times: user 24 ms, sys: 32 ms, total: 56 ms
# Wall time: 56.3 ms

x.nbytes / 0.056 / 1e9
# 1.4285714285714286

x = cupy.arange(50000000)

%time _ = device_to_host(x)
# CPU times: user 100 ms, sys: 148 ms, total: 248 ms
# Wall time: 246 ms

x.nbytes / 0.246 / 1e9
# 1.6260162601626018
```

Running this under prun I get the following results:

```
In [10]: %prun _ = device_to_host(x)

         152 function calls in 0.257 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.243    0.121    0.243    0.121 driver.py:286(safe_cuda_api_call)
        1    0.014    0.014    0.257    0.257 <string>:1(<module>)
        2    0.000    0.000    0.000    0.000 {method 'reduce' of 'numpy.ufunc' objects}
```

pentschev commented:
> It looks like even though we never put host-like things into self.device_buffer they can still appear to be there because that object looks at it's .slow attribute, which is self.host_buffer. 

It does indeed appear as the `.slow` attribute (which is what I expected), but that doesn't count towards the size of `self.device_buffer`. Could you elaborate on what's the problem with having that?

pentschev commented:
In general, I think this looks good. Also, I can reproduce the low bandwidth you mention. When I was doing some benchmarks on general serialization methods, I saw > 4GB/s, which I can't reproduce anymore (I am only sure I did because I had the timing saved on a notebook, which I overwrote now). I think this may be related to version of some library, I'll investigate this further today.

pentschev commented:
Ok, now I remember the reason for the bandwidth issue, it's host memory allocation:

```python
In [1]: import numpy as np

In [2]: from numba import cuda

In [3]: x = np.arange(1.25e8)

In [4]: d_x = cuda.to_device(x)

In [5]: %%timeit
   ...: y = d_x.copy_to_host()
   ...:
   ...:
477 ms ± 2.24 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %%timeit
   ...: dst = np.ones(x.shape)
   ...: y = d_x.copy_to_host(dst)
   ...:
   ...:
445 ms ± 207 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: dst = np.ones(x.shape)

In [8]: %%timeit
   ...: y = d_x.copy_to_host(dst)
   ...:
   ...:
154 ms ± 3.65 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The results above correspond to 2.1 GB/s, 2.25 GB/s and 6.5 GB/s, respectively.

I don't know how we could improve this, unless we kept a pool of host memory where we could copy the device data to, but this is something that would be difficult to maintain properly, since we don't want overload host memory with unused buffers. Any thoughts on this?

pentschev commented:
One more question, did all tests in `test_spill.py` work for you? The ones that spill to disk are raising `UnpicklingError`s for me.

mrocklin commented:
> One more question, did all tests in test_spill.py work for you?

Ah.  I haven't tried these yet.  I'll give them a shot in a bit.

> Any thoughts on this?

Numpy 1.17 is supposed to accelerate allocation of numpy arrays a bit and it was just released.  Maybe we should try that.

pentschev commented:
> Numpy 1.17 is supposed to accelerate allocation of numpy arrays a bit and it was just released. Maybe we should try that.

Just tried both NumPy 1.16.4 and 1.17.0, and the results are pretty consistent with those from my previous comment https://github.com/pentschev/dask-cuda/pull/1#issuecomment-516311982 in both cases. Unfortunately, no free speedup. :(

mrocklin commented:
>> One more question, did all tests in test_spill.py work for you?

> Ah. I haven't tried these yet. I'll give them a shot in a bit.

Nope.  These didn't work.  I tried looking at these tests but I haven't spent the time to really understand them.  The checks here are pretty nested, so at first glance I'm not sure what is going on.

pentschev commented:
> Nope. These didn't work. I tried looking at these tests but I haven't spent the time to really understand them. The checks here are pretty nested, so at first glance I'm not sure what is going on.

I assume you're talking about the `worker_assert` and related functions, right? These are basically only checking if there is anything on the `DeviceHostFile` (depending on the stage, we know what should be where, given the problem size). Regardless of those, the test still fails if we keep it very basic, for example, keeping only the following part of the test (deleting/commenting out everything that comes after that): https://github.com/rapidsai/dask-cuda/blob/branch-0.9/dask_cuda/tests/test_spill.py#L106-L111

That said, I'm pretty sure there's something that's not working as we expect to with deserialization.

mrocklin committed:
Build proper serialization for DeviceSerialized

mrocklin committed:
Avoid pause and kill in test_spill

Previously workers would reach near full storage on the data object and then
would pause or kill the worker.  For now we remove these behaviors.

I'm not sure why this was happening in the first place though.
Presumably the data should have been pushed to host almost immediately.

mrocklin commented:
I've rewritten the serialization manually (it's a nice thing to look at if you're curious about how to effectively serialize compound objects).  Things seem to mostly work in `test_spill.py`, although things would pause, as though the system wasn't choosing to push data from device to host.  I may leave this issue with you though if that's alright.  It seems like the serialization issues are gone.

Also, I noticed that I had old files lingering in my storage directory from previous versions.  These caused some confusion because I was trying to deserialize data that had been written during a previous iteration.  You may want to wipe your `dask-worker-space/storage` directory.

pentschev commented:
I can't reproduce the pausing you mentioned @mrocklin, either before or after your changes to `test_spill.py`. I suspect you may have some local configuration that is overriding the ones set in the tests.

Apart from that, CuPy tests pass, but deserializing cuDF fails here (and thus files don't get removed). Did you also run the cuDF tests or just CuPy ones? I'm wondering if the problem is only here or not. This is what I see:

```python-traceback
ERROR    asyncio:base_events.py:1608 Future exception was never retrieved
                                     future: <Future finished exception=ValueError('strides is incompatible with shape of requested array and size of buffer')>
Traceback (most recent call last):
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
    yielded = next(result)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/worker.py", line 1161, in delete_data
    self.release_key(key)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/worker.py", line 2016, in release_key
    if key in self.data and key not in self.dep_state:
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/_collections_abc.py", line 666, in __contains__
    self[key]
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/dask_cuda/device_host_file.py", line 149, in __getitem__
    return self.device_buffer[key]
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 70, in __getitem__
    return self.slow_to_fast(key)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 57, in slow_to_fast
    value = self.slow[key]
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/func.py", line 39, in __getitem__
    return self.load(self.d[key])
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 70, in __getitem__
    return self.slow_to_fast(key)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 57, in slow_to_fast
    value = self.slow[key]
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/func.py", line 39, in __getitem__
    return self.load(self.d[key])
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 392, in deserialize_bytes
    return deserialize(header, frames)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 190, in deserialize
    return loads(header, frames)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 56, in dask_loads
    return loads(header, frames)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/dask_cuda/device_host_file.py", line 73, in _
    part = deserialize(sub_header, frames[start:stop])
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 190, in deserialize
    return loads(header, frames)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 56, in dask_loads
    return loads(header, frames)
  File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/numpy.py", line 99, in deserialize_numpy_ndarray
    header["shape"], dtype=dt, buffer=frames[0], strides=header["strides"]
ValueError: strides is incompatible with shape of requested array and size of buffer
```

Any ideas what could be the cause for this?

mrocklin commented:
I would have to investigate to see.  Could I ask you to try this out
first?  I would start by serializing and deserializing a cudf dataframe,
then add in host_to_device/device_to_host , then add in
serialize/deserialize of the result, and so on.

On Wed, Jul 31, 2019 at 7:03 AM Peter Andreas Entschev <
notifications@github.com> wrote:

> I can't reproduce the pausing you mentioned @mrocklin
> <https://github.com/mrocklin>, either before or after your changes to
> test_spill.py. I suspect you may have some local configuration that is
> overriding the ones set in the tests.
>
> Apart from that, CuPy tests pass, but deserializing cuDF fails here (and
> thus files don't get removed). Did you also run the cuDF tests or just CuPy
> ones? I'm wondering if the problem is only here or not. This is what I see:
>
> ERROR    asyncio:base_events.py:1608 Future exception was never retrieved
>                                      future: <Future finished exception=ValueError('strides is incompatible with shape of requested array and size of buffer')>
> Traceback (most recent call last):
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
>     yielded = next(result)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/worker.py", line 1161, in delete_data
>     self.release_key(key)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/worker.py", line 2016, in release_key
>     if key in self.data and key not in self.dep_state:
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/_collections_abc.py", line 666, in __contains__
>     self[key]
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/dask_cuda/device_host_file.py", line 149, in __getitem__
>     return self.device_buffer[key]
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 70, in __getitem__
>     return self.slow_to_fast(key)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 57, in slow_to_fast
>     value = self.slow[key]
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/func.py", line 39, in __getitem__
>     return self.load(self.d[key])
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 70, in __getitem__
>     return self.slow_to_fast(key)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/buffer.py", line 57, in slow_to_fast
>     value = self.slow[key]
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/zict/func.py", line 39, in __getitem__
>     return self.load(self.d[key])
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 392, in deserialize_bytes
>     return deserialize(header, frames)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 190, in deserialize
>     return loads(header, frames)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 56, in dask_loads
>     return loads(header, frames)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/dask_cuda/device_host_file.py", line 73, in _
>     part = deserialize(sub_header, frames[start:stop])
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 190, in deserialize
>     return loads(header, frames)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 56, in dask_loads
>     return loads(header, frames)
>   File "/home/nfs/pentschev/miniconda3/envs/rapids-nightly-0.9/lib/python3.7/site-packages/distributed/protocol/numpy.py", line 99, in deserialize_numpy_ndarray
>     header["shape"], dtype=dt, buffer=frames[0], strides=header["strides"]ValueError: strides is incompatible with shape of requested array and size of buffer
>
> Any ideas what could be the cause for this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pentschev/dask-cuda/pull/1?email_source=notifications&email_token=AACKZTACAGBJNEBRLUTS7RDQCGLR7A5CNFSM4IHW3TOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3HLGMA#issuecomment-516862768>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACKZTDQSKDYSFVMAWAUER3QCGLR7ANCNFSM4IHW3TOA>
> .
>

pentschev commented:
Sure, I can also investigate further, I was wondering mostly if you immediately had an idea in mind. Just to confirm did you run cuDF tests?

mrocklin commented:
I ran all of the tests for DHF.  I ran some of the tests in test_spill, but
kept getting tripped up by the pause behavior.

On Wed, Jul 31, 2019 at 7:42 AM Peter Andreas Entschev <
notifications@github.com> wrote:

> Sure, I can also investigate further, I was wondering mostly if you
> immediately had an idea in mind. Just to confirm did you run cuDF tests?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pentschev/dask-cuda/pull/1?email_source=notifications&email_token=AACKZTHVTM47MMZPEETHSDDQCGQFXA5CNFSM4IHW3TOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3HPKFI#issuecomment-516879637>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACKZTCTG3BVEGQAU35SGSTQCGQFXANCNFSM4IHW3TOA>
> .
>

pentschev commented:
That's weird. I would check local configurations, I haven't ever got any of those tests pausing, not before, and still not with your latest changes.

pentschev commented:
Thanks @mrocklin for this PR, I managed to fix the issue I mentioned earlier, and everything seems to work now. Woohoo!!!

I can't figure out how to push to your branch, so I'm merging this in and pushing my changes to the original branch, and we can continue discussion in the original repo.
